core/src/main/java/org/apache/metamodel/query/parser/FromItemParser.java

```diff
@@ -136,16 +169,17 @@ private FromItem parseJoinItem(final String itemToken) {
             // rightPart = "b.id"
             final String rightPart = onClause.substring(indexOfEquals + 1).trim();
 
-            leftOn[i] = findSelectItem(leftPart, leftSide, rightSide);
-            rightOn[i] = findSelectItem(rightPart, leftSide, rightSide);
+            leftOn[i] = findSelectItem(leftPart, fromItems.toArray(new FromItem[fromItems.size()]), rightSide);
+            rightOn[i] = findSelectItem(rightPart, fromItems.toArray(new FromItem[fromItems.size()]), rightSide);
         }
-
-        final FromItem result = new FromItem(joinType, leftSide, rightSide, leftOn, rightOn);
+        
+        final FromItem leftItem = (leftFromItem != null) ? leftFromItem : leftSide;
+        final FromItem result = new FromItem(joinType, leftItem, rightSide, leftOn, rightOn);
         result.setQuery(_query);
         return result;
     }
 
-    private SelectItem findSelectItem(String token, FromItem leftSide, FromItem rightSide) {
+    private SelectItem findSelectItem(String token, FromItem[] leftSides, FromItem rightSide) {
```

Comment:
Actually, wouldn't it almost be better to just pass in the whole Set here instead of an array + an instance of FromItem?